src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,477 @@
+          case chargingVehicle @ ChargingVehicle(vehicle, _, _, station, _, _) if chargingVehicle.status == Waiting =>
+            log.debug(
+              s"Vehicle $vehicle is moved to waiting line at $tick in station $station, with {}/{} vehicles connected and {} in waiting line",
+              station.connectedVehicles.size,
+              station.zone.numChargers,
+              station.waitingLineVehicles.size
+            )
+            log.debug(s"connected vehicles: ${station.connectedVehicles.keys.mkString(",")}")
+            log.debug(s"waiting vehicles: ${station.waitingLineVehicles.keys.mkString(",")}")
+            sender ! WaitingInLine(tick, vehicle.id)
+          case chargingVehicle: ChargingVehicle =>
+            handleStartCharging(tick, chargingVehicle)
+        }
+      } else {
+        sender ! Failure(
+          new RuntimeException(
+            s"$vehicle is not a BEV/PHEV vehicle. Request sent by agent ${sender.path.name}"
+          )
+        )
+      }
+
+    case ChargingUnplugRequest(tick, vehicle, vehicleManager) =>
+      log.debug(s"ChargingUnplugRequest received for vehicle $vehicle from plug ${vehicle.stall} at $tick")
+      val physicalBounds = obtainPowerPhysicalBounds(tick, None)
+      val chargingNetwork = chargingNetworkMap(vehicleManager)
+      chargingNetwork.lookupVehicle(vehicle.id) match {
+        case Some(chargingVehicle) =>
+          val prevStartTime = chargingVehicle.latestChargingCycle.get.startTime
+          val startTime = Math.min(tick, prevStartTime)
+          val endTime = Math.max(tick, prevStartTime)
+          val duration = endTime - startTime
+          val (chargeDurationAtTick, energyToChargeAtTick) = dispatchEnergy(duration, chargingVehicle, physicalBounds)
+          chargingVehicle.processChargingCycle(startTime, energyToChargeAtTick, chargeDurationAtTick)
+          handleEndCharging(tick, chargingVehicle, Some(sender))
+        case _ =>
+          log.debug(s"Vehicle $vehicle is already disconnected at $tick")
+          sender ! UnhandledVehicle(tick, vehicle.id)
+      }
+
+    case Finish =>
+      chargingNetworkMap.foreach(_._2.clear())
+      chargingNetworkMap.clear()
+      powerController.close()
+      context.children.foreach(_ ! Finish)
+      context.stop(self)
+  }
+
+  /**
+    * Connect the vehicle
+    * @param tick current time
+    * @param chargingVehicle charging vehicle information
+    */
+  private def handleStartCharging(tick: Int, chargingVehicle: ChargingVehicle): Unit = {
+    val ChargingVehicle(vehicle, vehicleManager, _, _, _, theSender) = chargingVehicle
+    log.debug(s"Starting charging for vehicle $vehicle at $tick")
+    val physicalBounds = obtainPowerPhysicalBounds(tick, None)
+    vehicle.connectToChargingPoint(tick)
+    theSender ! StartingRefuelSession(tick, vehicle.id)
+    handleStartChargingHelper(tick, chargingVehicle)
+    val (chargeDurationAtTick, energyToChargeAtTick) =
```

Comment:
Everything after this seems a bit duplicated - shouldn't this be added and then let the PlanningTimeOutTrigger handle the below? Unless this is to handle the scenario from the time it starts UNTIL the next planning - which seems to be more forward calculation than reverse calculation? Maybe that is why you do the handleRefueling at the start of the planning - to handle this scenario? If so, then this is rather confusing and I still argue that you run a high risk of double refueling for 0 duration. This whole thing should be put into the same logic flow so it can only do the fuel dance in one place at one time. It seems kind of like doing a while loop where you have to prime it - correct. In this actor logic it is too disjointed and needs rewritten for clarity IMHO - lmk if that makes sense - or if I am totally off base.

Reply:
Yes it is convoluted to avoid double refueling or not refueling vehicles. The code is set with fallbacks. I tried to make it easier, but I didn't find any easier way to do so